File: leetcode/200.number-of-islands/200.number-of-islands-dfs.go
//DFS

func numIslands(grid [][]byte) int {
	if len(grid) == 0 || len(grid[0]) == 0 {
		return 0
	}

	landCount := 0
	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid[i]); j++ {
			if grid[i][j] == '1' {
				landCount = landCount + 1
				dfs(&grid, i, j)
			}
		}
	}
	return landCount
}

func dfs(grid *[][]byte, i int, j int) {
	m := len(*grid)
	n := len((*grid)[0])

	if i < m && j < n && i >= 0 && j >= 0 && (*grid)[i][j] == '1' {
		(*grid)[i][j] = '0' //visited
		dfs(grid, i-1, j)
		dfs(grid, i+1, j)
		dfs(grid, i, j-1)
		dfs(grid, i, j+1)
	} else {
		return
	}
}

File: leetcode-py/200.number-of-islands-dfs.py
from typing import List

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if len(grid) ==0 or len(grid[0]) == 0:
            return 0

        islands_count = 0

        def dfs(i, j):
            #base base: out of range or already visited. Notice OR
            if i<0 or i>=len(grid) or j<0 or j>=len(grid[0]) or grid[i][j] == "0":
                return


            grid[i][j] = "0" # mark visited

            dfs(i-1, j) #above
            dfs(i+1, j) #below
            dfs(i, j-1) #left
            dfs(i, j+1) #right


        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == "1": # only non-visited
                    islands_count +=1
                    dfs(i, j)

        return islands_count
